Confusing error message wording for moving out of non-copy slices.

https://is.gd/iH1SwT

``` rust
let x = vec![String::from("foo")].into_boxed_slice();
let y = x[0];

// error: cannot move out of type `[std::string::String]`, a non-copy fixed-size array [--explain E0508]
```

It refers to the slice as a fixed-sized array. This can also be reproduced with a `Vec<String>` by [indexing `(*x)[0]`](https://is.gd/rK0ddS) instead of [indexing `x[0]`](https://is.gd/5RW9Qy).

The error message today is better, but not ideal. `array` should probably be `slice`.

```
error[E0508]: cannot move out of type `[std::string::String]`, a non-copy array
 --> test.rs:3:9
  |
3 | let y = x[0];
  |         ^^^^
  |         |
  |         help: consider using a reference instead `&x[0]`
  |         cannot move out of here

error: aborting due to previous error(s)
```

--- a/src/librustc_borrowck/borrowck/gather_loans/move_error.rs
+++ b/src/librustc_borrowck/borrowck/gather_loans/move_error.rs
@@ -153,20 +153,19 @@ fn report_cannot_move_out_of<'a, 'tcx>(bccx: &BorrowckCtxt<'a, 'tcx>,
         }
 
         Categorization::Interior(ref b, mc::InteriorElement(ik)) => {
-            match (&b.ty.sty, ik) {
-                (&ty::TySlice(..), _) |
-                (_, Kind::Index) => {
-                    let mut err = struct_span_err!(bccx, move_from.span, E0508,
-                                                   "cannot move out of type `{}`, \
-                                                    a non-copy array",
-                                                   b.ty);
-                    err.span_label(move_from.span, "cannot move out of here");
-                    err
-                }
-                (_, Kind::Pattern) => {
+            let type_name = match (&b.ty.sty, ik) {
+                (&ty::TyArray(_, _), Kind::Index) => "array",
+                (&ty::TySlice(_), _) => "slice",
+                _ => {
                     span_bug!(move_from.span, "this path should not cause illegal move");
-                }
-            }
+                },
+            };
+            let mut err = struct_span_err!(bccx, move_from.span, E0508,
+                                           "cannot move out of type `{}`, \
+                                            a non-copy {}",
+                                           b.ty, type_name);
+            err.span_label(move_from.span, "cannot move out of here");
+            err
         }
 
         Categorization::Downcast(ref b, _) |

--- a/src/test/compile-fail/issue-12567.rs
+++ b/src/test/compile-fail/issue-12567.rs
@@ -15,12 +15,12 @@ fn match_vecs<'a, T>(l1: &'a [T], l2: &'a [T]) {
         (&[], &[]) => println!("both empty"),
         (&[], &[hd, ..]) | (&[hd, ..], &[])
             => println!("one empty"),
-        //~^^ ERROR: cannot move out of type `[T]`, a non-copy array
-        //~^^^ ERROR: cannot move out of type `[T]`, a non-copy array
+        //~^^ ERROR: cannot move out of type `[T]`, a non-copy slice
+        //~^^^ ERROR: cannot move out of type `[T]`, a non-copy slice
         (&[hd1, ..], &[hd2, ..])
             => println!("both nonempty"),
-        //~^^ ERROR: cannot move out of type `[T]`, a non-copy array
-        //~^^^ ERROR: cannot move out of type `[T]`, a non-copy array
+        //~^^ ERROR: cannot move out of type `[T]`, a non-copy slice
+        //~^^^ ERROR: cannot move out of type `[T]`, a non-copy slice
     }
 }
 

--- a/src/test/compile-fail/move-out-of-array-1.rs
+++ b/src/test/compile-fail/move-out-of-array-1.rs
@@ -24,5 +24,5 @@ fn main() {
 }
 
 fn foo(a: [D; 4], i: usize) -> D {
-    a[i] //~ ERROR cannot move out of type `[D; 4]`
+    a[i] //~ ERROR cannot move out of type `[D; 4]`, a non-copy array
 }

--- a/src/test/compile-fail/move-out-of-slice-1.rs
+++ b/src/test/compile-fail/move-out-of-slice-1.rs
@@ -15,7 +15,7 @@ struct A;
 fn main() {
     let a: Box<[A]> = Box::new([A]);
     match a {
-        box [a] => {}, //~ ERROR cannot move out of type `[A]`
+        box [a] => {}, //~ ERROR cannot move out of type `[A]`, a non-copy slice
         _ => {}
     }
 }

